File: src/h264/intra-pcm.ts
/** Pixel reconstruction used only by the expensive I_PCM fallback path. */
import { MBFlag, QUANTISER_SCALE } from "../mpeg2/constants.ts";
import type { Picture } from "../mpeg2/headers.ts";
import type { Macroblock } from "../mpeg2/macroblock.ts";
import { idct8 } from "./chroma.ts";
import type { PcmMacroblockSamples } from "./mb.ts";
import { fieldDctToFrameTargets, GRAY_DC, intraTargets } from "./quant.ts";

function sample(value: number): number {
  return Math.max(0, Math.min(255, Math.round(value)));
}

function mismatchControl(coeff: Float64Array): void {
  let sum = 0;
  for (let i = 0; i < 64; i++) {
    const value = Math.max(-2048, Math.min(2047, Math.trunc(coeff[i]!)));
    coeff[i] = value;
    sum += value;
  }
  if ((sum & 1) === 0) coeff[63] = Math.trunc(coeff[63]!) ^ 1;
}

/** Reconstruct one MPEG-2 intra macroblock as planar 4:2:0 samples. */
export function reconstructIntraPcm(
  mb: Macroblock,
  pic: Picture,
): PcmMacroblockSamples {
  if (mb.skipped || (mb.flags & MBFlag.INTRA) === 0) {
    throw new Error(
      "I_PCM reconstruction requires a coded MPEG-2 intra macroblock",
    );
  }
  const quantiserScale =
    QUANTISER_SCALE[pic.coding.qScaleType]![mb.quantiserScaleCode]!;
  const coeff = Array.from({ length: 6 }, () => new Float64Array(64));
  for (let block = 0; block < 6; block++) {
    const levels = mb.blocks[block];
    if (!levels) continue;
    const matrix = block < 4 ? pic.quant.intra : pic.quant.chromaIntra;
    const target = coeff[block]!;
    intraTargets(
      levels,
      matrix,
      quantiserScale,
      pic.coding.intraDcPrecision,
      target,
    );
    target[0] = target[0]! + GRAY_DC;
    mismatchControl(target);
  }

  // Field-DCT luma blocks carry alternating lines. Convert them into ordinary
  // upper/lower frame blocks before the inverse transform.
  if (mb.dctType === 1) {
    const converted = Array.from({ length: 4 }, () => new Float64Array(64));
    fieldDctToFrameTargets(coeff[0]!, coeff[2]!, converted[0]!, converted[2]!);
    fieldDctToFrameTargets(coeff[1]!, coeff[3]!, converted[1]!, converted[3]!);
    for (let block = 0; block < 4; block++) coeff[block] = converted[block]!;
  }

  const spatial = coeff.map(() => new Float64Array(64));
  for (let block = 0; block < 6; block++) idct8(coeff[block]!, spatial[block]!);

  const luma = new Uint8Array(256);
  for (let block = 0; block < 4; block++) {
    const x0 = (block & 1) * 8;
    const y0 = (block >> 1) * 8;
    for (let y = 0; y < 8; y++) {
      for (let x = 0; x < 8; x++) {
        luma[(y0 + y) * 16 + x0 + x] = sample(spatial[block]![y * 8 + x]!);
      }
    }
  }
  const plane = (block: number) =>
    Uint8Array.from(spatial[block]!, (value) => sample(value));
  return { luma, cb: plane(4), cr: plane(5) };
}

File: crates/mpeg2toh264/src/h264/intra_pcm.rs
//! Pixel reconstruction used only by the expensive I_PCM fallback path.

use crate::error::{bail, Result};
use crate::h264::chroma::idct8;
use crate::h264::mb::PcmMacroblockSamples;
use crate::h264::quant::{field_dct_to_frame_targets, intra_targets, FLAT_PREDICTION_DC};
use crate::mpeg2::constants::QUANTISER_SCALE;
use crate::mpeg2::headers::Picture;
use crate::mpeg2::macroblock::Macroblock;
use crate::round_half_up;

fn sample(value: f32) -> u8 {
    round_half_up(value as f64).clamp(0.0, 255.0) as u8
}

/// H.262 clause 7.4.4: the reconstructed coefficients are clipped and their sum
/// forced odd, so that different IDCT implementations cannot drift apart.
fn mismatch_control(coeff: &mut [f32; 64]) {
    let mut sum = 0i64;
    for value in coeff.iter_mut() {
        let clamped = value.trunc().clamp(-2048.0, 2047.0);
        *value = clamped;
        sum += clamped as i64;
    }
    if sum & 1 == 0 {
        coeff[63] = ((coeff[63].trunc() as i32) ^ 1) as f32;
    }
}

/// Reconstruct one MPEG-2 intra macroblock as planar 4:2:0 samples.
pub fn reconstruct_intra_pcm(mb: &Macroblock, pic: &Picture) -> Result<PcmMacroblockSamples> {
    if mb.skipped || !mb.is_intra() {
        bail!("I_PCM reconstruction requires a coded MPEG-2 intra macroblock");
    }
    let quantiser_scale =
        QUANTISER_SCALE[pic.coding.q_scale_type][mb.quantiser_scale_code as usize];
    let mut coeff = [[0.0f32; 64]; 6];
    for block in 0..6 {
        let Some(levels) = mb.block(block) else {
            continue;
        };
        let matrix = if block < 4 {
            &pic.quant.intra
        } else {
            &pic.quant.chroma_intra
        };
        let target = &mut coeff[block];
        intra_targets(
            levels,
            matrix,
            quantiser_scale,
            pic.coding.intra_dc_precision,
            target,
        );
        target[0] += FLAT_PREDICTION_DC;
        mismatch_control(target);
    }

    // Field-DCT luma blocks carry alternating lines. Convert them into ordinary
    // upper/lower frame blocks before the inverse transform.
    if mb.dct_type == 1 {
        let mut converted = [[0.0f32; 64]; 4];
        {
            let (upper, lower) = converted.split_at_mut(2);
            field_dct_to_frame_targets(&coeff[0], &coeff[2], &mut upper[0], &mut lower[0]);
            field_dct_to_frame_targets(&coeff[1], &coeff[3], &mut upper[1], &mut lower[1]);
        }
        coeff[..4].copy_from_slice(&converted);
    }

    let mut spatial = [[0.0f32; 64]; 6];
    let mut tmp = [0.0f32; 64];
    for block in 0..6 {
        idct8(&coeff[block], &mut spatial[block], &mut tmp);
    }

    let mut luma = [0u8; 256];
    for block in 0..4 {
        let x0 = (block & 1) * 8;
        let y0 = (block >> 1) * 8;
        for y in 0..8 {
            for x in 0..8 {
                luma[(y0 + y) * 16 + x0 + x] = sample(spatial[block][y * 8 + x]);
            }
        }
    }
    let mut cb = [0u8; 64];
    let mut cr = [0u8; 64];
    for i in 0..64 {
        cb[i] = sample(spatial[4][i]);
        cr[i] = sample(spatial[5][i]);
    }
    Ok(PcmMacroblockSamples { luma, cb, cr })
}
